sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py

```diff
@@ -249,6 +275,24 @@ def fused_lloyd_single_step(
         # End of outer loop. By now min_idx and min_sample_pseudo_inertia
         # contains the expected values.
 
+        # NB: this check can't be moved at the top at the kernel, because if a work item
+        # exits early with a `return` it will never reach the barriers, thus causing a
+        # deadlock. Early returns are only possible when there are no barriers within
+        # the remaining set of instructions for the running kernel.
+        if sample_idx >= n_samples:
+            return
+
+        if with_sample_weight:
+            weight = sample_weight[sample_idx]
+
+        if return_inertia:
+            inertia = min_sample_pseudo_inertia
+            if compute_exact_inertia:
+                inertia = X_l2_norm + (two * min_sample_pseudo_inertia)
+            if with_sample_weight:
+                inertia = inertia * weight
+            per_sample_inertia[sample_idx] = inertia
```

Comment:
>  I think we can remove the first branch

The weight is reused when updating centers later in the kernel, that's why the branch is used here

Reply:
Copying a reply in a past comment about `with_sample_weight` use in general:

> My hope (here and for all the other occurences of this strategy in this PR) are that since `with_sample_weight` is a constant (not a direct input to the kernel) the compiler will detect the dead branch and ignore it when compiling, I've read that this is a common optimization optimizers can do, and also I've not seen differences in the benchmarks before/after this PR.